apps/prairielearn/src/models/course-instances.ts

```diff
@@ -0,0 +1,52 @@
+import { callValidatedOneRow } from '@prairielearn/postgres';
+import { z } from 'zod';
+import { IdSchema } from '../lib/db-types';
+import { idsEqual } from '../lib/id';
+
+const CourseInstanceAuthzSchema = z.object({
+  course_instances: z.array(
+    z.object({
+      short_name: z.string().nullable(),
+      long_name: z.string().nullable(),
+      id: IdSchema,
+      formatted_start_date: z.string(),
+      formatted_end_date: z.string(),
+      has_course_instance_permission_view: z.boolean(),
+    }),
+  ),
+});
+
+export async function selectAuthorizedCourseInstancesForCourse({
+  course_id,
+  user_id,
+  authn_user_id,
+  is_administrator,
+}: {
+  course_id: string;
+  user_id: string;
+  authn_user_id: string;
+  is_administrator: boolean;
+}) {
+  const { course_instances: authnCourseInstances } = await callValidatedOneRow(
+    'course_instances_with_staff_access',
+    [authn_user_id, is_administrator, course_id],
+    CourseInstanceAuthzSchema,
+  );
+
+  if (idsEqual(user_id, authn_user_id)) {
```

Comment:
In the course function we have `authz_data_overrides` -- why do we check ID equality for course instances but use `authz_data_overrides` for courses? This is presumably some holdover from old code that I'm just not tracking.

Reply:
If there is `authz_data_overrides` it means we have an effective user, role, or similar. In such cases, switching courses is disabled, and only the current course is allowed. However, in such cases switching instances is still allowed, and in that case permissions are an intersection of those from the authorized user and those of the authenticated user.

Reply:
Ah, excellent, thanks for the clear explanation!